### PR TITLE
Mc us10 12

### PR DESCRIPTION
**Pull request by Mariana-21**
## Spec/Features ##
* user login_spec created

## Routes ## 
* Get route for new method for register
* post route for create method in register
*Get route for show method for profiles

## Controllers ## 
* show controller created in profile
* "new" controller created in registration
* create controller created in registration

## Views ##
* registration new form 
* profile show page created

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    elsif email_exists?
+      flash[:error] = "Email has already been taken"
+      render :new
+    else
+      flash[:notice] = "#{missing_params(user_params).join(", ")} can't be blank, please try again."
+      render :new
+    end
+  end
+
+  private
```
Comment (foleymichelle9) [inline app/controllers/users_controller.rb:22]: Nice use of private methods!

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,10 +1,14 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
 
-  helper_method :cart
+  helper_method :cart, :current_user
```
Comment (perryr16) [inline app/controllers/application_controller.rb:4]: Good use of helper method

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      session[:user_id] = @user.id
+      flash[:success] = "Welcome #{@user.name}! You are now registered and logged in!"
+      redirect_to "/profile"
+    elsif email_exists?
+      flash[:error] = "Email has already been taken"
+      render :new
+    else
+      flash[:notice] = "#{missing_params(user_params).join(", ")} can't be blank, please try again."
+      render :new
+    end
```
Comment (dborski) [inline app/controllers/users_controller.rb:19]: This is a clever way of displaying error messages. Like it

File: spec/features/user/login_spec.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      fill_in :address,	with: @user_info2[:address]
+      fill_in :city, with: @user_info2[:city]
+      fill_in :state,	with: @user_info2[:state]
+      # Missing zip field
+      fill_in :email,	with: @user_info2[:email]
+      fill_in :password, with: @user_info2[:password]
+      fill_in :confirm_password,	with: @user_info2[:password]
+      click_button "Submit"
+
+      expect(current_path).to eq("/register")
```
Comment (foleymichelle9) [inline spec/features/user/login_spec.rb:89]: These tests look great!

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    @user = User.new(user_params)
+    if @user.save
+      session[:user_id] = @user.id
+      flash[:success] = "Welcome #{@user.name}! You are now registered and logged in!"
+      redirect_to "/profile"
+    elsif email_exists?
+      flash[:error] = "Email has already been taken"
+      render :new
+    else
+      flash[:notice] = "#{missing_params(user_params).join(", ")} can't be blank, please try again."
```
Comment (perryr16) [inline app/controllers/users_controller.rb:17]: Nice use of private method here

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,33 @@
+<h1>New Registration Form</h1>
+
+
+
+<center>
+  <%= form_tag "/register" do %>
```
Comment (perryr16) [inline app/views/users/new.html.erb:6]: We should use a partial in the future

**Comment by foleymichelle9** [comment]
Looks great! I approve!!

File: spec/features/user/login_spec.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
+require 'rails_helper'
+
+RSpec.describe "As a visitor", type: :feature do 
+  describe "When I click on the 'register' link in the nav bar" do
+    before(:each) do
```
Comment (perryr16) [inline spec/features/user/login_spec.rb:5]: could use factory_bot for these

File: db/migrate/20200527212017_remove_password_confirmation_from_users.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class RemovePasswordConfirmationFromUsers < ActiveRecord::Migration[5.1]
+  def change
+    remove_column :users, :password_confirmation
```
Comment (perryr16) [inline db/migrate/20200527212017_remove_password_confirmation_from_users.rb:3]: nice, password confirmation shouldn't be in the resource

**Pull request merged**